src/main/scala/org/apache/spark/metrics/source/custom/AccumulatorMetricsSource.scala

```diff
@@ -0,0 +1,54 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.apache.spark.metrics.source.custom
+
+import java.{lang => jl}
+
+import com.codahale.metrics.{Gauge, Meter, MetricRegistry}
+import org.apache.spark.metrics.source.Source
+import org.apache.spark.util.AccumulatorV2
+
+import scala.collection.mutable
+
+/**
+  * Accumulator-based metric source
+  * <p><p>
+  * Usage:
+  * <pre>{@code
+  * val metricsSource = new AccumulatorMetricsSource("prefix")
+  * val metric = spark.sparkContext.longAccumulator("metric-name")
+  * metricsSource.registerAccumulator(metric, metric.name.getOrElse("UNNAMED_ACCUMULATOR"))
+  * SparkEnv.get.metricsSystem.registerSource(metricsSource)
+  * }</pre>
+  * @param prefix for metrics collected by this source
+  */
+class AccumulatorMetricsSource(prefix: String) extends Source {
+  override def sourceName: String = prefix
+  override def metricRegistry: MetricRegistry = registry
+
+  private val registry = new MetricRegistry
+  private val gauges = mutable.Map[String, Long]()
+  private val meters = mutable.Map[String, Meter]()
+
+  def registerAccumulator(acc: AccumulatorV2[jl.Long, jl.Long], name: String) {
+    gauges.put(name, 0L)
+    meters.put(name, metricRegistry.meter(name + "-rate"))
```

Comment:
Is it necessary to maintain the mutable maps here? I think we could instantiate the gauge and the meter directly and reference them in the gauge's `getValue` method without having to look them up in a map.